### PR TITLE
Service proxy stub

### LINKED ISSUE #40: Make service proxy available for testing
It's very common to have the necessity to unit test a module which relies on a service proxy, which it receives at runtime with the method `getDirectServiceProxy`, typically by dependency injection. In this case, the only way is the manually build a stub of the service proxy in order to do assertions about the calls to the proxed service.
It would be useful to have such a stub (builder?) available inside the lib and requirable by the unit test.

```js
const {ServiceProxyStub} = require('@mia-platform/custom-plugin-lib')

const MyService = require('...')

const serviceStub = new ServiceProxyStub()
const myService = new MyService(serviceStub)

myServive.doStuff()
assertCallsToService(serviceStub)
```

The alternative is to make the `getDirectServiceProxy` method available outside the context of the `custom-plugin` so that the test can create an actual service proxy and then use libraries like `nock` to intercept the http calls (it's still stubbing at the http level).

```js
const nock = require('nock')

const {getDirectServiceProxy} = require('@mia-platform/custom-plugin-lib')

const MyService = require('...')

const service = getDirectServiceProxy('crud-service', {...})
const myService = new MyService(service)

setupInterceptors(nock)
myServive.doStuff()
assertInterceptors(nock)
```

### PR DESCRIPTION
**Pull request by fabio-percivaldi**
Resolves #40

#### Checklist

- [X] your branch will not cause merge conflict with `master`
- [X] run `npm test`
- [X] tests are included
- [X] the documentation is updated or intregrated accordingly with your changes
- [X] the code will follow the lint rules and style of the project
- [X] you are not committing extraneous files or sensible data

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -216,3 +216,8 @@ function initCustomServiceEnvironment(envSchema = defaultSchema) {
 }
 
 module.exports = initCustomServiceEnvironment
+module.exports.getDirectServiceProxy = getDirectlyServiceBuilderFromService
+module.exports.getServiceProxy = (microserviceGatewayServiceName) => {
+  this.MICROSERVICE_GATEWAY_SERVICE_NAME = microserviceGatewayServiceName
+  return getServiceBuilderFromService.bind(this)
+}
```
Comment (allevo) [inline index.js:223]:
```suggestion
module.exports.getServiceProxy = function (microserviceGatewayServiceName, baseOptions = {}) {
  return getServiceBuilderFromService(MICROSERVICE_GATEWAY_SERVICE_NAME, {}, baseOptions)
}
```

In this way we avoid to change `this` and the double function invocation

Comment (fabio-percivaldi) [reply index.js:223]: to do this we should change the `getServiceBuilderFromService(baseOptions = {})` signature or we can directly call the `serviceBuilder(MICROSERVICE_GATEWAY_SERVICE_NAME, {}, baseOptions)`

Comment (allevo) [reply index.js:223]: right, maybe the second way is better

**Comment by fabio-percivaldi** [comment]
> Miss the typescript definition. Could you add it?

i don't know typescript so i tried my best, in case i would appreciate your correction for the ts definition

**Comment by allevo** [comment]
> i don't know typescript so i tried my best, in case i would appreciate your correction for the ts definition

there's some test on types (see tests directory) that can help you to see if the definition is good

File: tests/types/index.ts [hunk 2]
```diff
@@ -220,3 +224,10 @@ a(async function (service) {
     return request.abortChain(200, { final: 'body' }, { some: 'other headers' })
   })
 })
+
+async function invokeProxies() {
+  const directServiceProxy = getDirectServiceProxy('service_name', { port:3000 })
```
Comment (allevo) [inline tests/types/index.ts:229]:
```suggestion
  const directServiceProxy = getDirectServiceProxy('service_name')
  const directServiceProxyWithOptions = getDirectServiceProxy('service_name', { port:3000 })
```

File: tests/types/index.ts [hunk 3]
```diff
@@ -220,3 +224,10 @@ a(async function (service) {
...
+  const serviceProxy = getServiceProxy('microservice-gateway', { port: 3000 })
```
Comment (allevo) [inline tests/types/index.ts:230]:
```suggestion
  const serviceProxy = getServiceProxy('microservice-gateway')
  const serviceProxyWithOptions = getServiceProxy('microservice-gateway', { port: 3000 })
```

**Pull request merged**